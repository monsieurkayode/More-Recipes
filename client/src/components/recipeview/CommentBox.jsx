import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'proptypes';

import { postReview } from '../../actions';
import validate from '../../utils/validateInput';

/**
 * @summary - CommentBox class declaration
 * @class CommentBox
 * @extends {Component}
 */
export class CommentBox extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf CommentBox
   */
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * Handle submit
   *
   * @method onSubmit
   *
   * @param {string} comment
   *
   * @returns {void}
   */
  onSubmit(comment) {
    const { id } = this.props.currentRecipe;
    const { reset } = this.props;
    this.props.postReview(id, comment)
      .then(() => reset());
  }

  /**
   * Render comment section
   *
   * @method renderCommentBox
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderCommentBox = (field) => {
    const { input } = field;
    return (
      <div className="input-field">
        <textarea
          {...input}
          className="materialize-textarea"
          id="reviews"
          type="text"
          placeholder="Add a comment.."
        />
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { handleSubmit } = this.props;
    return (
      <div id="comment-box" className="row">
        <div className="col l7 m8 s12 offset-l4 offset-m4">
          <form
            onSubmit={handleSubmit(this.onSubmit)}
          >
            <Field
              name="comment"
              component={this.renderCommentBox}
            />
            <button
              id="post-btn"
              type="submit"
              className="btn btn-flat right"
            >
              Post
            </button>
          </form>
        </div>
      </div>
    );
  }
}

CommentBox.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  postReview: PropTypes.func.isRequired,
  currentRecipe: PropTypes.shape({
    id: PropTypes.number
  }).isRequired,
  reset: PropTypes.func.isRequired
};

const mapStateToProps = ({ currentRecipe }) => ({ currentRecipe });

export default reduxForm({
  validate,
  form: 'PostCommentForm'
})(connect(mapStateToProps, { postReview })(CommentBox));
