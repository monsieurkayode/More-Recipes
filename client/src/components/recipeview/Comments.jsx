import React from 'react';
import PropTypes from 'proptypes';
import Avatar from 'react-avatar';
import { formatTimeFromNow } from '../../utils/timeFormat';

import pascalCase from '../../utils/pascalCase';

const Comments = ({ review }) => {
  let time = review.createdAt;
  time = formatTimeFromNow(time);
  return (
    <div className="review-container z-depth-2">
      <div className="review-body">
        <div className="user-avatar">
          <Avatar
            className="responsive-img"
            name={review.User.username}
            size={60}
            round
          />
        </div>
        <div className="review-content-block">
          <div>
            <p className="username">{pascalCase(review.User.username)}</p>
            <p className="review-time">{time}</p>
          </div>
          <div className="review-text">
            <p>{review.comment}</p>
          </div>
          <div className="review-controls">
            <i className="fa fa-pencil"> <span>Edit</span> </i>
            <i className="fa fa-trash"> <span>Delete</span></i>
          </div>
        </div>
      </div>
    </div>
  );
};

Comments.propTypes = {
  review: PropTypes.shape({
    comment: PropTypes.string,
    createdAt: PropTypes.string,
    User: PropTypes.shape({
      username: PropTypes.string
    })
  }).isRequired
};

export default Comments;
