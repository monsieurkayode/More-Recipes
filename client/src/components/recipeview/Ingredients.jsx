import React from 'react';
import PropTypes from 'proptypes';
import showdown from 'showdown';

showdown.setFlavor('github');

/**
 * Ingredients
 * @function Ingredients
 *
 * @param {object} props
 *
 * @return {JSX} JSX
 */
const Ingredients = (props) => {
  const converter = new showdown.Converter();
  return (
    <div className="col l3 m4 s12 offset-l1">
      <span className=""><h5>Ingredients</h5></span>
      <p className="divider" />
      <div
        id="ingredients"
        // eslint-disable-next-line
        dangerouslySetInnerHTML={
          { __html: converter.makeHtml(props.ingredients) }
        }
      />
    </div>
  );
};

Ingredients.defaultProps = {
  ingredients: ''
};

Ingredients.propTypes = {
  ingredients: PropTypes.string.isRequired
};
export default Ingredients;
