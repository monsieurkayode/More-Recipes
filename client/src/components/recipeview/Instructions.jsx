import React from 'react';
import PropTypes from 'proptypes';
import showdown from 'showdown';

showdown.setFlavor('github');

/**
 * Instructions
 * @function Instructions
 *
 * @param {object} props
 *
 * @return {JSX} JSX
 */
const Instructions = (props) => {
  const converter = new showdown.Converter();
  return (
    <div className="col l12 m12 s12">
      <h4>Instructions</h4>
      <div
        id="instructions"
        // eslint-disable-next-line
        dangerouslySetInnerHTML={
          { __html: converter.makeHtml(props.instructions) }
        }
      />
    </div>
  );
};

Instructions.defaultProps = {
  instructions: ''
};

Instructions.propTypes = {
  instructions: PropTypes.string.isRequired
};

export default Instructions;
