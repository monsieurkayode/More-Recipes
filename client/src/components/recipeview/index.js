import Comments from './Comments';
import Ingredients from './Ingredients';
import Instructions from './Instructions';
import CommentBox from './CommentBox';

export {
  Comments,
  Ingredients,
  Instructions,
  CommentBox
};
