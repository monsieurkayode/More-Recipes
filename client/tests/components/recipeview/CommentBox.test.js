/* eslint-disable no-undef */
import React from 'react';

import { CommentBox } from '../../../src/components/recipeview/CommentBox';
import data from '../../__mocks__/mockData';

const renderSpy = jest.spyOn(CommentBox.prototype, 'render');

const setup = () => {
  const props = {
    reset: jest.fn(),
    handleSubmit: jest.fn(),
    currentRecipe: data.fetchRecipesResponse.recipes[0],
    postReview: jest.fn().mockImplementation(() => Promise.resolve())
  };

  const wrapper = shallow(<CommentBox {...props} />);

  return {
    wrapper,
    props
  };
};

const { wrapper, props } = setup();

describe('CommentBox', () => {
  it('should call render method',
    (done) => {
      expect(renderSpy).toHaveBeenCalled();
      done();
    });

  it('should render correctly',
    (done) => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });

  it('should have a form with a textarea',
    (done) => {
      const form = wrapper.find('form');
      expect(form.exists()).toBe(true);
      done();
    });

  it('should have a submit button for the form',
    (done) => {
      const submitButton = wrapper.find('button');
      expect(submitButton.exists()).toBe(true);
      done();
    });

  it('should call renderCommentBox',
    (done) => {
      wrapper.instance().renderCommentBox({ input: null });
      done();
    });

  it('should call handleSubmit when the form is submitted',
    (done) => {
      const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'onSubmit');
      props.handleSubmit(wrapper.instance().onSubmit());
      expect(handleSubmitSpy).toHaveBeenCalled();
      done();
    });
});
