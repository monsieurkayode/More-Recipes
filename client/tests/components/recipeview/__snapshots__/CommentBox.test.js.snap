// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentBox should render correctly 1`] = `
<div
  className="row"
  id="comment-box"
>
  <div
    className="col l7 m8 s12 offset-l4 offset-m4"
  >
    <form>
      <Field
        component={[Function]}
        name="comment"
      />
      <button
        className="btn btn-flat right"
        id="post-btn"
        type="submit"
      >
        Post
      </button>
    </form>
  </div>
</div>
`;
